### DISCUSSION TITLE
oscillating training error

### DISCUSSION THREAD

**Discussion by negf** [deepmodeling/deepmd-kit#175] [Q&A]
![training_error](https://user-images.githubusercontent.com/13686746/73990585-f43b5680-498c-11ea-9b7e-7ecb1190970c.png)

Hi,

After the both training and test error decreased in the initial steps, both starts to oscillate later in the training (see attached figure).  I have ~ 60000  training sets, which is split into 4 sets with different number of atoms if this makes a different. The input is 

{
    "_comment": " model parameters",
    "model": {
        "type_map":     ["K"],
        "descriptor" :{
            "type":             "se_ar",
            "a" : {
                "sel":          [120],
                "rcut_smth":    9.0,
                "rcut":         10.0,
                "neuron":       [128,64,32,16],
                "resnet_dt":    false,
                "axis_neuron":  16,
                "seed":         1,
                "_comment":     " that's all"
            },
            "r": {
                "sel":          [120],
                "rcut_smth":    10.0,
                "rcut":         13.0,
                "neuron":       [128,64,32,16],
                "resnet_dt":    false,
                "seed":         1,
                "_comment":     " that's all"
            }
        },
        "fitting_net" : {
            "neuron":           [128,128,128,128],
            "resnet_dt":        false,
            "seed":             1,
            "_comment":         " that's all"
        },
        "_comment":     " that's all"
    },

    "learning_rate" :{
        "type":         "exp",
        "start_lr":     1E-4,
        "decay_steps":  100000,
        "decay_rate":   0.97,
        "_comment":     "that's all"
    },

    "loss" :{
        "start_pref_e": 0.01,
        "limit_pref_e": 1,
        "start_pref_f": 1000,
        "limit_pref_f": 1,
        "start_pref_v": 0,
        "limit_pref_v": 0,
        "_comment":     " that's all"
    },

    "_comment": " traing controls",
    "training" : {
        "systems":      ["./k111_wire3", "./k111_wire2", "./k111_wire3_2", "./k111_wire3_2_1", "./k111_wire3_1atom_chain_broken"],
        "set_prefix":   "set",
        "stop_batch":   25000000,
        "batch_size":   "auto",
        "seed":         1,

        "_comment": " display and restart",
        "_comment": " frequencies counted in batch",
        "disp_file":    "lcurve.out",
        "disp_freq":    100,
        "numb_test":    10,
        "save_freq":    1000,
        "save_ckpt":    "model.ckpt",
        "load_ckpt":    "model.ckpt",
        "disp_training":true,
        "time_training":true,
        "profiling":    false,
        "profiling_file":"timeline.json",
        "_comment":     "that's all"
    },

    "_comment":         "that's all"
}

(base)

Best,
Marius

**Comment by njzjz** [comment]
Please take the logarithm of both x-axis and y-axis and plot a new figure. Thanks!

**Comment by negf** [comment]
![log_training_error](https://user-images.githubusercontent.com/13686746/73992727-513a0b00-4993-11ea-93de-c92df3d5a0af.png)

Here the log plot

**Comment by jameswind** [comment]
The oscillating behavior should be due to the fact that the on-the-fly
training and testing errors are measured by the data randomly visited in
different systems. So the accuracy of the model is different for
different systems.
On the other hand, it seems that 1) the training error is large and 2)
there seems to be some overfitting since the testing accuracy is worse than
training.
For 1), could you also show the error of forces? I also find some places
that might improve the accuracy: "neuron": [128,64,32,16], could be changed
to, say, [20, 40, 80], since it is embedding a one-dimensional quantity (r)
onto a high-dimensional space; start_lr is typically 1e-3. Please see other
reference values at
https://github.com/deepmodeling/deepmd-kit/blob/master/examples/water/train/water_se_a.json
For 2), I feel that 60000 training data is a lot. Is there any reason why
overfitting happens?

**Comment by negf** [comment]
![force_log](https://user-images.githubusercontent.com/13686746/74122337-214a6c00-4c0e-11ea-860d-b4d2350dc368.png)
![enegery_log](https://user-images.githubusercontent.com/13686746/74122338-227b9900-4c0e-11ea-8f41-f84a3883f1a2.png)

I changed the setting following your suggestions, here are the force and energy errors. It did not really improve much on the oscillating behavior. I honestly  don't see the reason for overfitting. I think it got better now but difficult to say, though.
